src/view/components/microbit/MicrobitSimulator.tsx

```diff
@@ -0,0 +1,162 @@
+import * as React from "react";
+import CONSTANTS, { DEVICE_LIST_KEY } from "../../constants";
+import PlayLogo from "../../svgs/play_svg";
+import StopLogo from "../../svgs/stop_svg";
+import Dropdown from "../Dropdown";
+import ActionBar from "../simulator/ActionBar";
+import { MicrobitImage } from "./MicrobitImage";
+
+const initialLedState = [
+    [0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0],
+    [0, 0, 0, 0, 0],
+];
+
+interface vscode {
+    postMessage(message: any): void;
+}
+
+declare const vscode: vscode;
+
+const sendMessage = (type: string, state: any) => {
+    vscode.postMessage({ command: type, text: state });
+};
+interface IState {
+    active_editors: string[];
+    running_file: string;
+    leds: number[][];
+    play_button: boolean;
+    selected_file: string;
+}
+export class MicrobitSimulator extends React.Component<any, IState> {
+    constructor() {
+        super({});
+        this.state = {
+            leds: initialLedState,
+            play_button: false,
+            selected_file: "",
+            active_editors: [],
+            running_file: "",
+        };
+        this.onSelectBlur = this.onSelectBlur.bind(this);
+    }
+    handleMessage = (event: any): void => {
+        const message = event.data;
+        if (message.active_device !== DEVICE_LIST_KEY.MICROBIT) {
+            return;
+        }
+
+        switch (message.command) {
+            case "reset-state":
+                console.log("Reset the state");
+                this.setState({
+                    ...this.state,
+                    leds: initialLedState,
+                    play_button: false,
+                });
+                break;
+            case "set-state":
+                console.log("Setting the state");
+                this.setState({
+                    leds: message.state.leds,

```

Comment:
this one is missing ...this.state

Also, are they even necessary? I thought setState patches the fields that are given to it.

Reply:
Yea, they don't seem to be necessary. I usually don't use it unless i'm modifying an object inside the state and I want to keep the attributes the same.